Lib/fontTools/feaLib/builder.py

```diff
@@ -1270,46 +1270,91 @@ def add_subtable_break(self, location):
-    def __init__(self, font, location):
-        LookupBuilder.__init__(self, font, location, 'GPOS', 8)
-        self.rules = []  # (prefix, input, suffix, lookups)
-
+class ChainContextualBuilder(LookupBuilder):
     def equals(self, other):
         return (LookupBuilder.equals(self, other) and
                 self.rules == other.rules)
 
+    def buildLookup_(self, subtables):
+        return otl.buildLookup(subtables, self.lookupflag, self.markFilterSet,
+            optimize="chaining_format", builder=self)
+
     def build(self):
         subtables = []
         for (prefix, glyphs, suffix, lookups) in self.rules:
             if prefix == self.SUBTABLE_BREAK_:
                 continue
-            st = otTables.ChainContextPos()
+            st = self.newSubtable_()
             subtables.append(st)
             st.Format = 3
             self.setBacktrackCoverage_(prefix, st)
             self.setLookAheadCoverage_(suffix, st)
             self.setInputCoverage_(glyphs, st)
-
-            st.PosCount = 0
-            st.PosLookupRecord = []
             for sequenceIndex, lookupList in enumerate(lookups):
                 if lookupList is not None:
                     if not isinstance(lookupList, list):
                         # Can happen with synthesised lookups
                         lookupList = [ lookupList ]
                     for l in lookupList:
-                        st.PosCount += 1
                         if l.lookup_index is None:
                             raise FeatureLibError('Missing index of the specified '
-                                'lookup, might be a substitution lookup',
+                                f'lookup, might be a {self.other} lookup',
                                 self.location)
-                        rec = otTables.PosLookupRecord()
-                        rec.SequenceIndex = sequenceIndex
-                        rec.LookupListIndex = l.lookup_index
-                        st.PosLookupRecord.append(rec)
+                        self.addLookupRecord_(st, sequenceIndex, l.lookup_index)
         return self.buildLookup_(subtables)
 
+    def add_subtable_break(self, location):
+        self.rules.append((self.SUBTABLE_BREAK_, self.SUBTABLE_BREAK_,
+                           self.SUBTABLE_BREAK_, [self.SUBTABLE_BREAK_]))
+
+
+class ChainContextPosBuilder(ChainContextualBuilder):
+    def __init__(self, font, location):
+        LookupBuilder.__init__(self, font, location, 'GPOS', 8)
+        self.rules = []  # (prefix, input, suffix, lookups)
+
+    def newSubtable_(self):
+        st = otTables.ChainContextPos()
+        st.PosCount = 0
+        st.PosLookupRecord = []
+        return st
+
+    def newChainRuleSet_(self, st, index=None):
+        if not hasattr(st, "ChainPosRuleSetCount"):
+            st.ChainPosRuleSetCount = 0
+            st.ChainPosRuleSet = []
+        ruleset = otTables.ChainPosRuleSet()
+        st.ChainPosRuleSetCount += 1
+        if index:
```

Comment:
```suggestion
        if index is not None:
```
In case `index` can be zero (or `if index is None` and invert the condition).